[21047] Work package activities also display repository revisons

oliverguenther opened the pull request:
This commit adds preliminary support for showing aggregated regular work
package activities and revisions linked to that particular work package
(through commit message referencing).
### TODOS
- [x] APIv3 endpoint must receive a property to output `formatted_identifier` (e.g., Git SHA shortened to 7chars)
- [x] Activities are only shown when fully aggregated. Need to check for errors in the revisions collection retrieval
- [x] Can we introduce a more abstract `activity` template to build upon? `revisions_activity` and `user_activity` share quite a bit of HTML.
- [x] Unit tests
- [X] Integration tests
### Notes

Author of a commit is not always linked to a user, if the committers were not set. Currently only the author name is shown in that case. We need to identify whether users should be matched heuristically (e.g., by name or mail), or rather pessimistically.

Related work packages:
https://community.openproject.org/work_packages/15422
https://community.openproject.org/work_packages/21047

oliverguenther committed:
[21047] Work package activities also display repository revisons

This commit adds preliminary support for showing aggregated regular work
package activities and revisions linked to that particular work package
(through commit message referencing).

Related work packages:
https://community.openproject.org/work_packages/15422
https://community.openproject.org/work_packages/21047

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-repository-activity.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-repository-activity.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

oliverguenther commented on frontend/app/work_packages/tabs/revision-activity-directive.js:
```diff
@@ -0,0 +1,64 @@
+
+      scope.project = scope.workPackage.embedded.project;
+      scope.revision = scope.activity.props.identifier;
+      scope.formattedRevision = (scope.revision.length > 8 ? scope.revision.substr(0,8) : scope.revision);
```
Note: this is only a temporary fix to let me see the resulting revision date line more clearly. The formatted identifier will eventually be returned from the revisions endpoint.

floriank commented on frontend/app/helpers/path-helper.js:
```diff
@@ -76,6 +76,12 @@ module.exports = function() {
+    },
+    revisionPath: function(projectIdentifier, revisionIdentifier) {
+      return PathHelper.repositoryPath(projectIdentifier) + '/revision/' + revisionIdentifier;
+    },
```
not to bitch here, but shouldn't we use HAL here and infer the links from the api response for the project?

oliverguenther replied:
I already suggested to have this link returned from the revisions endpoint instead:https://github.com/opf/openproject/pull/3326

floriank commented on frontend/app/templates/work_packages/tabs/_revision_activity.html:
```diff
@@ -0,0 +1,22 @@
+         href="#{{ activityNo }}"></a>
+    </span>
+  </div>
+  <img class="avatar" ng-src="{{ userAvatar }}" alt="Avatar" title="{{userName}}" ng-if="userAvatar" />
```
Using `ng-if` with an image can flicker shortly in IE10 and IE11, as the broser attempts to load the image initialyy, better to wrap it into a `div` imho.

oliverguenther replied:
Thanks. I'll take those comments into account when moving this into its own templates, as your comments currently apply to [_user_activity.html](https://github.com/opf/openproject/blob/feature/repository_activity/frontend/app/templates/work_packages/tabs/_user_activity.html) as well (cf. todo in PR note)

floriank commented on frontend/app/templates/work_packages/tabs/_revision_activity.html:
```diff
@@ -0,0 +1,22 @@
+  <span class="user" ng-if="userActive"><a ng-href="{{ userPath(userId) }}" name="{{ currentAnchor }}" ng-bind="userName"></a></span>
+  <span class="user" ng-if="!userActive">{{ userName }}</span>
+  <span class="date">
+    <a href="{{ revisionPath }}" title="{{ revision }}">
```
use `ng-href` please.

floriank commented on frontend/app/work_packages/controllers/work-package-details-controller.js:
```diff
@@ -172,13 +174,53 @@ module.exports = function($scope,
   };
 
-  function displayedActivities(workPackage) {
+  $scope.matchesActivityType = function(activity, type) {
```
while this is a nice idea, it's not readable very well.

I coul image giving the underlying controller some information about the types and then define some more readable helper functions for the viewModel, i.e. `isRevisionType`

oliverguenther replied:
I'd prefer if we could wire a directive implicitly to a specific resource, but this seems reasonable. I will wrap the check in a helper, although I disagree with the readability issue.

floriank commented on frontend/app/work_packages/controllers/work-package-details-controller.js:
```diff
@@ -172,13 +174,53 @@ module.exports = function($scope,
+        aggregated = aggregated.concat(activity);
+      }
+
+      if (++totalActivities === 2) {
```
incrementing a variable in a condition. good lord. :gun:

oliverguenther replied:
I wholeheartedly disagree. I find this very concise and it does one specific task only.
I know Crockford's view on that and agree about potential readability issues with pre-increment in a more complex situation, but this isn't `while (*foo++ = *bar++);` ;-)

NobodysNightmare replied:
But `while (*foo++ = *bar++);` is still a quite elegant way to write a segfault ^^

floriank commented on frontend/app/templates/work_packages/tabs/_revision_activity.html:
```diff
@@ -0,0 +1,22 @@
+    <span ui-sref="work-packages.list.details.activity.details({ activity_no: activityNo })">
+      <a name="{{ activityNo }}"
+         ng-bind="'#' + activityNo"
+         href="#{{ activityNo }}"></a>
```
`ng-href` ?

oliverguenther committed:
Use ng-switch to determine activity type

oliverguenther committed:
Move activities out of tabs

This moves activity templates and directives in their own
namespaces / module.
Also adds a single activity entry for overview and activity,
which itself selects the correct directive for the given activity
object.

oliverguenther committed:
Remove superfluous path

Revision path is now returned from API endpoint and thus
can be removed from the path helper.

oliverguenther committed:
Add missing href to resource

oliverguenther committed:
Add unit test for revision activity

NobodysNightmare commented:
regarding

> Time of commit is not relative

This made you wonder, because all other times are displayed as absolute times.

Reading the acceptance criteria:

> the time is displayed in the known format (e.g. "11 months ago")

I would say that the criteria should read as "do it as it is done for the activities".

I can well imagine that the format of choice will become "relative", but this can well be done separately for both activites and revisions...

edit: Since the relevant WP is an EPIC I'd propose to do it in a separate PR, but ask for clarification anyway.

oliverguenther commented:
I agree that this should be moved in a separate PR. I've also extracted it from the revision epic, as it also applies to regular activities:

https://community.openproject.org/work_packages/21227

NobodysNightmare commented on config/locales/js-en.yml:
```diff
@@ -82,6 +83,7 @@ en:
     label_collapse: "Collapse"
     label_collapsed: "collapsed"
     label_collapse_all: "Collapse all"
+    label_committed_at: "committed revision %{revision}"
```
The key and value do not match.
I would expect a value of _committed revision %{revision}_ **_at**_

oliverguenther replied:
You're right, I removed `at` from the value simply because the whole label becomes the link text, and at shouldn't be part of it. 

It could also be that the link text is still subject to change, e.g., so that only the `revision %{revision}` part is clickable. Do you have any input on that?

NobodysNightmare replied:
What you are describing sounds like a hell of a localization problem. Because we won't be able to just do:

``` html
<a href="">{{ t('label_created') }}</a> {{ t('label_at') }} {{ getTheDate() }}
```

This is not translatable very well... the best thing would be something along the lines of `committed revision %{revision} at {date}`, because this offers the best translatability. But we lose our fine grained linkability, especially since we do not want HTML in our translations.

oliverguenther replied:
Is it acceptible to inject HTML through the interpolations? I've seen this happen sometimes in the backend, and while it's not pretty, I'd rather go down that road than to touch these labels soon again because the order just won't work in all languages for `<a href="foo">:label_committed_at</a> :label_at <a href="bar">{{ date }}</a>`.

NobodysNightmare replied:
best of both worlds proposal by @ulferts (I am writing what he explained):

``` yml
label_my_full_sentence: "%{committed_revision_link} at %{date}"
a_placeholder_for_my_full_sentence: "committed revision %{revision}"
```

and then you build your matroska (ruby example):

``` ruby
sentence = I18n.t('label_my_full_sentence',
                  committed_revision_link: I18n.t('a_placeholder_for_my_full_sentence',
                    revision: my_revision).make_fancy_link_out_of_that,
                  date: my_date)
```

The translators will have to make sure that both translations match up well, but at least they have the chance to build working sentences....

edit: Yes, I think HTML in interpolations is perfectly fine, as long as the HTML does not originate from another translation ;-)

oliverguenther replied:
@NobodysNightmare While this change wasn't trivial due to date being a directive itself and thus needing to compile that whole string, I've adapted the suggested nested localization in https://github.com/oliverguenther/openproject/commit/1e11547af365d21a7bc8e824f85bf08e6a366370

NobodysNightmare commented on frontend/app/templates/work_packages/tabs/activity.html:
```diff
@@ -10,12 +10,14 @@
+            <activity-entry work-package="workPackage"
+                           activity="activity"
+                           activity-no="activityNo"
+                           is-initial="isInitial"
```
Hint: `isInitial` is calculated as `activityNo == 1` and influences the correct display of the first WP activity ("was created" and does not display changes)...

I know it is unlikely, but I could imagine some commit referencing a WP before it is created:
- because of bad timestamps in VCS
- because of unintentional mentioning ("see GitHub PR #123 on that unrelated repository")

That could make a revision the "initial" activity, couldn't it?

oliverguenther replied:
That's a good point, and something I'd have to check with how the journal is initially filled when a work package is created. 

I don't think the second one would work, though, or can you mention a work package prior to its creation using the `#id` notation? Even if you mention some WP unintentionally, the created journal entry still would precede it.

NobodysNightmare replied:
> I don't think the second one would work, though, or can you mention a work package prior to its creation using the `#id` notation? Even if you mention some WP unintentionally, the created journal entry still would precede it.

This is my fear. That I could unintentionally reference it prior its creation and later when the WP is created it will be linked to that commit and thereby become the "initial" activity as of the current implementation.

~~If this was not possible (I don't know for sure), the current implementation (`activityNo == 1`) may be sufficient...~~ edit: there is still case 1

oliverguenther replied:
I'll check how the current implementation handles that, otherwise we will have to work around it.

oliverguenther replied:
You were goddamn right with your fear. I've committed the changeset before the WP was created, and it is indeed the first journal. Since I sort the journals by date, I've addressed it in the frontend (i.e., in this PR).

oliverguenther committed:
Correct for wrong isInitial

oliverguenther committed:
Revision integration tests

oliverguenther committed:
Better localized revision link using dynamically compiled date and href

oliverguenther commented:
@NobodysNightmare I've accounted for the localization remark and [added a workaround for initial activities that are not of type `Activity`](https://github.com/oliverguenther/openproject/commit/ccd86a4b75c3192a8533aa6461c34db0a486d9c8). I'll assign this to you once more since you're already familiar with the rest of the PR.

oliverguenther committed:
Remove revision unit test example that requires compile

NobodysNightmare commented on frontend/app/work_packages/controllers/work-package-details-controller.js:
```diff
@@ -172,13 +174,78 @@ module.exports = function($scope,
+
+
+    // Type must be Activity
+    if (type !== 'Activity') {
```
`&& type !== 'Activity::Comment'` -.-

I always thought that this differentiation sucks... but it is there and apparently it is used in the angular client as well...

Noticed that this is neccessary, because of a first journal of mine which is tagged as `Activity::Comment`.

oliverguenther replied:
Oh, alright. Why exactly can a first activity journal be a comment? I assume that happens if you add a note while wp creation? Then the check below must also be extended.

Would it suffice to check that the type starts with `Activity`? Or is that type likely to extend to other, non-initial journals?

oliverguenther committed:
Test for Activity and Activity::Comment, which may also be initial

NobodysNightmare commented:
LGTM, but waiting for travis :+1:
